frame/assets/src/lib.rs

```diff
@@ -247,19 +256,594 @@ decl_error! {
+		/// minimum for the asset, then the amount burned is increased to take it to zero.
+		///
+		/// Weight: `O(1)`
+		/// Modes: Post-existence of `who`
+		#[weight = T::WeightInfo::burn()]
+		fn burn(origin,
+			#[compact] id: T::AssetId,
+			who: <T::Lookup as StaticLookup>::Source,
+			#[compact] amount: T::Balance
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			let who = T::Lookup::lookup(who)?;
+
+			Asset::<T>::try_mutate(id, |maybe_details| {
+				let d = maybe_details.as_mut().ok_or(Error::<T>::Unknown)?;
+				ensure!(&origin == &d.admin, Error::<T>::NoPermission);
+
+				let burned = Account::<T>::try_mutate_exists(
+					id,
+					&who,
+					|maybe_account| -> Result<T::Balance, DispatchError> {
+						let mut account = maybe_account.take().ok_or(Error::<T>::BalanceZero)?;
+						let mut burned = amount.min(account.balance);
+						account.balance -= burned;
+						*maybe_account = if account.balance < d.min_balance {
+							burned += account.balance;
+							Self::dead_account(&who, d, account.is_zombie);
+							None
+						} else {
+							Some(account)
+						};
+						Ok(burned)
+					}
+				)?;
+
+				d.supply = d.supply.saturating_sub(burned);
+
+				Self::deposit_event(RawEvent::Burned(id, who, burned));
+				Ok(())
+			})
+		}
+
+		/// Move some assets from the sender account to another.
+		///
+		/// Origin must be Signed.
+		///
+		/// - `id`: The identifier of the asset to have some amount transferred.
+		/// - `target`: The account to be credited.
+		/// - `amount`: The amount by which the sender's balance of assets should be reduced and
+		/// `target`'s balance increased. The amount actually transferred may be slightly greater in
+		/// the case that the transfer would otherwise take the sender balance above zero but below
+		/// the minimum balance. Must be greater than zero.
+		///
+		/// Emits `Transferred` with the actual amount transferred. If this takes the source balance
+		/// to below the minimum for the asset, then the amount transferred is increased to take it
+		/// to zero.
+		///
+		/// Weight: `O(1)`
+		/// Modes: Pre-existence of `target`; Post-existence of sender; Prior & post zombie-status
+		/// of sender.
+		#[weight = T::WeightInfo::transfer()]
+		fn transfer(origin,
+			#[compact] id: T::AssetId,
+			target: <T::Lookup as StaticLookup>::Source,
+			#[compact] amount: T::Balance
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			ensure!(!amount.is_zero(), Error::<T>::AmountZero);
+
+			let mut origin_account = Account::<T>::get(id, &origin);
+			ensure!(!origin_account.is_frozen, Error::<T>::Frozen);
+			origin_account.balance = origin_account.balance.checked_sub(&amount)
+				.ok_or(Error::<T>::BalanceLow)?;
+
+			let dest = T::Lookup::lookup(target)?;
+			Asset::<T>::try_mutate(id, |maybe_details| {
+				let d = maybe_details.as_mut().ok_or(Error::<T>::Unknown)?;
+
+				if dest == origin {
+					return Ok(())
+				}
```

Comment:
This can be above the try mutate to avoid the storage read

Reply:
No. Passing an invalid asset ID is always an error, even if you only want to transfer some assets to yourself. If the asset class doesn't exist then, you can't possibly transfer them, even to yourself.